### ISSUE TITLE
[BUG] Core 15x slower reading blobs than Framework

### ISSUE THREAD

**Issue by OracPrime** [Azure/azure-sdk-for-net#17149]
**Describe the bug**
I am seeing the v12 Blobs library taking 45 seconds to read a blob via netcore that the same library can read in 3 seconds in Net4.7

**Expected behavior**
I would expect core to be at least as fast as framework

**Actual behavior (include Exception or Stack Trace)**
Framework is 15x faster than core.  Framework appears to have requests active in parallel, core seems to serialise 16kb chunks.

**To Reproduce**
Run the following code (obviously providing a connection string for a premium blob) in both core and framework

```using System;
using System.Diagnostics;
using Azure.Storage.Blobs;
using Azure.Storage.Blobs.Models;

 

namespace BlobTest
{
  class Program
  {
    static void Main(string[] args)
    {
      Stopwatch sw = Stopwatch.StartNew();
      BlobClient blobClient = new BlobClient(new Uri( <insert premium blob address here> ));
      int toRead = 500 * 1024 * 1024;
      byte[] byteBuffer = new byte[toRead];
      Azure.Response<BlobDownloadInfo> response = blobClient.Download(new Azure.HttpRange(0,toRead));
      using (BlobDownloadInfo result = response.Value)
      {
        int followOns = 0;
        long read = result.ContentLength;
        long read2 = result.Content.Read(byteBuffer, 0, byteBuffer.Length);
        while (read2 < read)
        {
            ++followOns;
            long read3 = result.Content.Read(byteBuffer, (int)read2, byteBuffer.Length - (int)read2);
            read2 += read3;
            Debug.Assert(read3 > 0, "Failed to read blob");
        }
        sw.Stop();
        Console.WriteLine($"Read {toRead} in {sw.ElapsedMilliseconds}ms {followOns} followOns");
      }
    }
  }
}
```

**Environment:**
 - Azure.Storage.Blobs 12.6.0
 - Comparing Net Framework 4.7.2 and dotnetcore 3.1.301
 - VS2019 16.6.2
Running on Azure VM D4sv3 with accelerated networking connecting to a premium block storage account in the same region (West Europe)

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team best able to assist.

**Comment by jaschrep-msft** [comment]
Hi @OracPrime. I have a couple clarifying questions.

I assume you are running the net472 app and netcoreapp3.1 app on the same VM and not across different ones? If you are running on different VMs, is the VM running the netcore3.1 app still Windows? .NET Core is known to have differing performance across different operating systems.

- Are you running this in debug mode through visual studio or as a release-compiled console app? Interactive debug mode will have significant performance implications.
- `Framework appears to have requests active in parallel, core seems to serialise 16kb chunks.` Can you go into greater detail on this?
  - The specific `Download()` method you are calling does not touch the code path that will parallelize requests, while various `DownloadTo()` methods will, provided the right configuration arguments. I'd like to better understand the parallelization you're encountering on this specific code snippet you've provided.
  - Are the 16kb chunks in netcore on your .NET Read() methods or are there individual GET HTTP requests going out that request a 16KB range in the headers? This is of interest to me, as SDK defaults are in the megabyte range, and I don't see you specifying chunk sizes anywhere.

A general question alongside the above: does switching from `Download()` to `DownloadToAsync()` fix this situation? `Download()` is much closer to single web requests, only handling optimized retries for you, whereas `DownloadTo()` manages whole downloads to a destination for you, with configurable performance tweaks. I specifically call out `DownloadToAsync()` over `DownloadTo()`, as there is significantly better async performance on this method for large downloads as of 12.6.0.

**Comment by OracPrime** [comment]
Both versions running on the same VM, running release build outside the IDE.

My comments came from looking at Fiddler and/or Insomnia when it was running.  Fiddler was showing on its diagram multiple parallel streams for the framework case, but only one for netcore.  Netcore disaplyed a series of continuation messages indicating additional 16kb chunks had arrived.

I'll investigate the Download/DownloadTo options and get back to you.

**Comment by OracPrime** [comment]
The blob we have is potentially huge, and we are just reading part of it (to memory, not disk).  None of the DownloadTo methods allow for a partial download as far as I can see?

**Comment by OracPrime** [comment]
I've run some more tests on my test data, which is about 250MB.  The fastest I can achieve is 1.3s using 64 parallel copies of this function to read a slice:
```
    private static async Task MakeSubSliceTask(long blobOffset, int byteOffset, int length)
    {
      Azure.Response<BlobDownloadInfo> response = blobClient.Download(new Azure.HttpRange(blobOffset, length));

      using (BlobDownloadInfo result = response.Value)
      {
        int followOns = 0;
        int readFragment = 0;
        int totalRead = 0;
        while (totalRead < result.ContentLength)
        {
          ++followOns;
          byteOffset += readFragment;
          length -= readFragment;
          readFragment = await result.Content.ReadAsync(byteBuffer, byteOffset, length);
          totalRead += readFragment;
          Debug.Assert(readFragment > 0, "Failed to read blob");
        }
      }
    }
```
If I create an appropriately size MemoryStream and use DownloadTo then it takes 2.1s, if I use DownloadToAsync it is 2.0s

For all three of these techniques, speed is pretty much identical for core or framework (perhaps 10% faster in core).  It's only the single Download call in my original post which has the core/framework disparity.

The best of all worlds would seem to be if DownloadTo and DownloadToAsync had an overload which took a range, for when you don't need the entire blob (or as in our case, where different slices of the blob are being read by different k8s pods)

**Comment by jaschrep-msft** [comment]
I did not run this in a high-speed VM/Storage environment, but I ran your sample copy/pasted with my own 256 MB blob subbed in and this was the output produced:

```
netcore3.1: Read 524288000 in 95917ms 16413 followOns
net472:     Read 524288000 in 73497ms 16413 followOns
```

I don't have the most consistent speeds in my environment, I'm more concerned with the underlying operations going on, but I'm nowhere near the 15x difference you encounter. Additionally, I ran this while monitoring fiddler. I only saw a single HTTP request for each Framework and Core, and they were nearly identical, minus some cosmetic differences. So I'm not sure I understand what you mean when you mention the parallel streams in framework versus the serialized 16kb chunks in core. I'd appreciate very fine-grained detail on this.

**Comment by OracPrime** [comment]
It looks as though you're getting the slow way both ways.  My net performance for the same size data is only 2.1 seconds.  I'll reach out to you directly to progress this further

**Comment by OracPrime** [comment]
Github privacy levels have beaten me.  Feel free to contact me at my FirstName.LastName@willistowerswatson.com and we can set up a higher bandwidth discussion.  I'm in UK timezone.

**Comment by amishra-dev** [comment]
@jaschrep-msft, James will the fixed made for the perf improvement recently help David (@OracPrime )?

**Comment by jaschrep-msft** [comment]
No that is a separate issue. #17305 has been filed to work around this issue. Meanwhile, I understand David is managing their own parallel Download() requests.

**Comment by jaschrep-msft** [comment]
Closing to favor #17305 as the solution to this situation.
